tool/tsh/common/mcp.go

```diff
@@ -0,0 +1,238 @@
+	dryRun              bool
+}
+
+func newMCPDBCommand(parent *kingpin.CmdClause) *mcpDBCommand {
+	cmd := &mcpDBCommand{
+		CmdClause: parent.Command("db", "Start a local MCP server for database access"),
+	}
+
+	cmd.Flag("db-user", "Database user to log in as.").Short('u').StringVar(&cmd.databaseUser)
+	cmd.Flag("db-name", "Database name to log in to.").Short('n').StringVar(&cmd.databaseName)
+	cmd.Flag("labels", labelHelp).StringVar(&cmd.labels)
+	cmd.Flag("query", queryHelp).StringVar(&cmd.predicateExpression)
+	return cmd
+}
+
+func (c *mcpDBCommand) run(cf *CLIConf) error {
+	logger, _, err := logutils.Initialize(logutils.Config{
+		Severity: slog.LevelInfo.String(),
+		Format:   mcpLogFormat,
+		Output:   mcpLogOutput,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	registry := defaultDBMCPRegistry
+	if cf.databaseMCPRegistryOverride != nil {
+		registry = cf.databaseMCPRegistryOverride
+	}
+
+	tc, err := makeClient(cf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Avoid any input request on the command execution. This is required,
+	// otherwise the MCP clients will be stuck waiting for a response.
+	tc.NonInteractive = false
+
+	sc, err := newSharedDatabaseExecClient(cf, tc)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	dbs, err := c.getDatabases(cf.Context, sc)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	server := dbmcp.NewRootServer()
+	allDatabases, closeLocalProxies, err := c.prepareDatabases(cf.Context, registry, dbs, logger, tc, sc.profile, server)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer closeLocalProxies()
+
+	for protocol, newServerFunc := range registry {
+		databases := allDatabases[protocol]
+		if len(databases) == 0 {
+			continue
+		}
+
+		srv, err := newServerFunc(cf.Context, &dbmcp.NewServerConfig{
+			RootServer: server,
+			Databases:  databases,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer srv.Close(cf.Context)
+	}
+
+	return trace.Wrap(server.ServeStdio(cf.Context, cf.Stdin(), cf.Stdout()))
+}
+
+// closeLocalProxyFunc function used to close local proxy listeners.
+type closeLocalProxyFunc func() error
+
+// prepareDatabases based on the available MCP servers, initialize the database
+// local proxy and generate the MCP database.
+func (c *mcpDBCommand) prepareDatabases(
+	ctx context.Context,
+	registry dbmcp.Registry,
+	dbs []types.Database,
+	logger *slog.Logger,
+	tc *client.TeleportClient,
+	profile *client.ProfileStatus,
+	server *dbmcp.RootServer,
+) (map[string][]*dbmcp.Database, closeLocalProxyFunc, error) {
+	var (
+		dbsPerProtocol = make(map[string][]*dbmcp.Database)
+		closeFuncs     []closeLocalProxyFunc
+	)
+
+	for _, db := range dbs {
+		if !registry.IsSupported(db.GetProtocol()) {
+			logger.InfoContext(ctx, "database protocol unsupported, skipping it", "database", db.GetName(), "protocol", db.GetProtocol())
+			continue
+		}
+
+		route := tlsca.RouteToDatabase{
+			ServiceName: db.GetName(),
+			Protocol:    db.GetProtocol(),
+		}
```

Comment:
DatabaseUser and other routes? I think this will only work for certain db protocols like postgres where we overwrite the route based on handshake.